Feature : merge databases program

louib opened the pull request:
## Description

Adding a util script for merging 2 database files.

## Motivation and Context

I wanted to be able to merge automatically database files from the command line. Since the merging functionality is now part of the project, I wrote a script to leverage it.

## How Has This Been Tested?
Mainly from the command line:
1. 2 identical database files, the first file remains unchanged.
2. Some entries are in file2 but not in file1. They are present in file1 after merging.
3. file2 modified an entry of file1. file1 gets its entry updated.

That's how far I went with testing the conflict resolution itself, as it was handled with https://github.com/keepassxreboot/keepassxc/pull/47

I don't think this change affects other areas of the code.

It might be a good idea to document the util scripts available, but i'm not really sure were to add the new documentation (in the source code, in the wiki?).

Also, having problems running the tests, that's why it's not checked yet...

## Types of changes
- :negative_squared_cross_mark: Bug fix
- ✅ New feature
- :negative_squared_cross_mark: Breaking change

## Checklist
- ✅  I have read the **CONTRIBUTING** document. [REQUIRED]
- ✅  My code follows the code style of this project. [REQUIRED]
- ✅ All new and existing tests passed. [REQUIRED]
- ✅ My change requires a change to the documentation.
- :negative_squared_cross_mark: I have updated the documentation accordingly.
- :negative_squared_cross_mark: I have added tests to cover my changes.

droidmonkey commented on .gitignore:
```diff
@@ -1,2 +1,4 @@
 CMakeLists.txt.*
 build*/
+*.swp
+cmake/
```
Why was .gitignore changed? Please revert this.

louib replied:
@droidmonkey my bad, will revert this.

droidmonkey commented:
Your code is failing to compile on TravisCI. Also, this is not a script but a compiled executable program.

TheZ3ro commented:
TravisCI is failing because there are some error in your file.

https://travis-ci.org/keepassxreboot/keepassxc/jobs/189663430#L1054-L1179
They are noted from line 1054 to line 1179, you should fix those error.

louib commented:
@droidmonkey reverted the .gitignore `cmake` line.
@TheZ3ro the tests are passing now. Looks like they were easier to run on linux than on MacOS.

TheZ3ro commented:
The code is good, I would refrain from entering the password as a command line parameter but kdbx-extract also do this (https://github.com/keepassxreboot/keepassxc/blob/develop/utils/kdbx-extract.cpp)

Just a note, the code you wrote assume that the db1 and db2 use the same password. Should be nice accepting a different password for the second database.

Also, what does `*.swp` stand for in the `.gitignore` file?

JosefVitu commented:
@TheZ3ro  `*.swp` is a swap file, generated for example by vim.

@louib Maybe I'd suggest using a global `.gitignore` file for your needs (see `man gitignore(5)`).

TheZ3ro commented:
@JosefVitu yeah, I was asking what was the need to insert `*.swp` in the .gitignore file

louib commented:
@JosefVitu thanks for the tip, I'll do that instead and revert the changes to the .gitignore!
@TheZ3ro I'm not super comfortable with passing the passphrase as an argument either... I could use the `pw-stdin` option as in the main program. sounds good?

TheZ3ro commented:
If you didn't read this:
> Just a note, the code you wrote assume that the db1 and db2 use the same password. Should be nice accepting a different password for the second database.

I've seen the password reading from std, nice.
I will test this in the next few days

louib commented:
@TheZ3ro Added the second password for the second database. Could be nice to add an option to use the same password for the 2 databases. Something like `--same-password` maybe.

Prompting the user for a password could also be cool (with the option `-p`?). Unfortunately, I did not find any secure and cross-platform way in C++ to prompt the user from the terminal for a password. If you has any suggestion, I'd be glad to look at it further, especially if other util scripts are to be written.

There is duplication in the code, since we read both password files the same way. It could be a good idea to extract this code into functions to avoid duplication. I think some code in the `main.cpp` file, reading the password from stdin, is also similar. Not sure where I would move these util functions though...

TheZ3ro commented:
> @TheZ3ro Added the second password for the second database. Could be nice to add an option to use the same password for the 2 databases. Something like --same-password maybe.
 
I was thinking about inserting an empty password as first or second but the empty string is a valid database password, so we need to add a `--same-password` or `-s`. I agree typing 2 time a long password is useless and boring.

> Prompting the user for a password could also be cool (with the option -p?). Unfortunately, I did not find any secure and cross-platform way in C++ to prompt the user from the terminal for a password. If you has any suggestion, I'd be glad to look at it further, especially if other util scripts are to be written.

Do you mean a graphic prompt? If yes, this are CLI script so I don't think it's useful

> There is duplication in the code, since we read both password files the same way. It could be a good idea to extract this code into functions to avoid duplication. I think some code in the main.cpp file, reading the password from stdin, is also similar. Not sure where I would move these util functions though...

Yes we should make a new file but I don't know where, maybe @debfx can point us a right way to do this

louib commented:
> we need to add a --same-password or -s. I agree typing 2 time a long password is useless and boring.

I agree. I will add that option! I think I will also use `QCommandLineParser` to have a better help message and handle the options more easily.

> Do you mean a graphic prompt? If yes, this are CLI script so I don't think it's useful

No, I was thinking about a terminal prompt, so the user could type the password interactively. As I said, this raises security concern on it's own, so I decided not implement it, relying instead on stdin.

TheZ3ro commented:
Ok great, seems fine to me! :smile:

louib commented:
@TheZ3ro I Switched to QCommandLineParser, which will give the following output:

```bash
louib@Desktop $ ./utils/kdbx-merge 
Usage: ./utils/kdbx-merge [options] database1 database2
Merge 2 KeePassXC database files.

Options:
  -h, --help           Displays this help.
  -s, --same-password  use the same password for both database files.

Arguments:
  database1            path of the database to merge into.
  database2            path of the database to merge from.
```

I renamed the utility `kdbx-merge`, to stay consistent with `kdbx-extract`. As you can see, I also implemented the `-s` option, for reusing the password for both databases.

TheZ3ro commented:
I will test this in the next few days ;)

TheZ3ro commented on utils/CMakeLists.txt:
```diff
@@ -25,6 +25,16 @@ target_link_libraries(kdbx-extract
+add_executable(kdbx-merge kdbx-merge.cpp)
+target_link_libraries(kdbx-merge
+                      keepassx_core
+                      ${MHD_LIBRARIES}
```
Here MHD lib are useless, also I would remove Qt5::Concurrent and Qt5::Widgets.
It works fine with just:
```
                      keepassx_core
                      Qt5::Core
                      ${GCRYPT_LIBRARIES}
                      ${ZLIB_LIBRARIES}
```

TheZ3ro commented:
Changes approved. Test ok, code review ok.

Ping @droidmonkey
